### PR TITLE
feat: Support casting Duration to String in ISO 8601 format

### PR DESCRIPTION
**Pull request by kanenorman**
Fixes: #21964

This PR addresses two related issues:

## 1. Implement `Duration` to `String` cast

Casting a `Duration` series to `String` now produces ISO 8601 duration strings:

```python
>>> import polars as pl
>>> pl.Series([1]).cast(pl.Duration("ns")).cast(pl.String)
shape: (1,)
Series: '' [str]
[
        "PT0.000000001S"
]
```

## 2. Fix duration sub-second formatting 

Sub-second components are now decomposed per unit rather than collapsed into nanoseconds, as [proposed in the comments](https://github.com/pola-rs/polars/issues/21964#issuecomment-2761263304)

```python
>>> import polars as pl
>>> from datetime import timedelta
>>> pl.DataFrame(
...     {
...         "d": [
...             timedelta(weeks=1),
...             timedelta(days=1),
...             timedelta(hours=1),
...             timedelta(minutes=1),
...             timedelta(seconds=1),
...             timedelta(milliseconds=1),
...             timedelta(microseconds=1),
...             timedelta(),
...             timedelta(
...                 weeks=1,
...                 days=1,
...                 hours=1,
...                 minutes=1,
...                 seconds=1,
...                 milliseconds=1,
...                 microseconds=1,
...             ),
...         ]
...     },
...     schema={'d': pl.Duration(time_unit="ns")},
... )
shape: (9, 1)
┌─────────────────────┐
│ d                   │
│ ---                 │
│ duration[ns]        │
╞═════════════════════╡
│ 7d                  │
│ 1d                  │
│ 1h                  │
│ 1m                  │
│ 1s                  │
│ 1ms                 │
│ 1μs                 │
│ 0ns                 │
│ 8d 1h 1m 1s 1ms 1μs │
└─────────────────────┘
```

Happy to split these into separate PRs if preferred. 
Also, I am a first time contributor, I have read the guidelines and agree to follow the instructions that were outlined.

First-time contributors must adhere to the following rules, or your PR(s) will be closed:
- You must post a screenshot of you successfully running the test suite (`make test`),
  locally on your machine (not the CI). ✅
image
- You may not have more than one open PR at a time. ✅

Note that if you used AI to generate code there are additional requirements you
must fulfill for your PR to be reviewed. See the contribution guidelines for
details, afterwards you can use the following template if you wish:

  1. I used AI to generate the test cases in `test_duration_cast_to_string`, iron out logic/edge cases in `fmt_duration_string`, draft this PR
  3. I confirm that I have reviewed all changes myself, and I believe they are
  relevant and correct.

### CODE REVIEW

**Review by nameexhaustion — CHANGES_REQUESTED** [review]
I don't think we need to be adding this amount of new code to make the casting work, since we already have the implementation to convert them to strings - 

```python
>>> pl.Series([(1<<27)]).cast(pl.Duration('ms')).dt.to_string(format='polars')
shape: (1,)
Series: '' [str]
[
        "1d 13h 16m 57s 728ms"
]
```

Could you find that implementation and re-work this PR to make `cast(String)` dispatch to that instead? It should just work afterwards.

**Comment by kanenorman** [comment]
Hi @nameexhaustion,

Thanks for the feedback.

I can reduce the size & scope of this PR and focus only on casting `Duration` to `String`.
I will drop the formatting related commits for now.

Before proceeding, do you have a preference for the output format?
I currently use ISO to match `Datetime` and be consistent, but I notice your comment is referring to the "polars" format. I can switch if needed.

Should the behavior be:

```python
>>> import polars as pl
>>> pl.Series([1]).cast(pl.Duration("ns")).cast(pl.String)
shape: (1,)
Series: '' [str]
[
    "PT0.000000001S"
]
```

or:

```python
>>> import polars as pl
>>> pl.Series([1]).cast(pl.Duration("ns")).cast(pl.String)
shape: (1,)
Series: '' [str]
[
    "1ns"
]
```

Happy to adjust based on your preference.

**Comment by nameexhaustion** [comment]
Let's go for ISO format

File: crates/polars-core/src/series/implementations/duration.rs [hunk 1]
```diff
@@ -435,7 +435,10 @@ impl SeriesTrait for SeriesWrap<DurationChunked> {
     }
 
     fn cast(&self, dtype: &DataType, cast_options: CastOptions) -> PolarsResult<Series> {
-        self.0.cast_with_options(dtype, cast_options)
+        match dtype {
+            DataType::String => Ok(self.0.to_string("iso")?.into_series()),
+            _ => self.0.cast_with_options(dtype, cast_options),
```
Comment (nameexhaustion) [inline crates/polars-core/src/series/implementations/duration.rs]: Could we move this 1 level lower to https://github.com/pola-rs/polars/blob/3ee3f1baa8deedc712e45d822cafc62eef2a3af7/crates/polars-core/src/chunked_array/logical/duration.rs#L29-L38

File: crates/polars-plan/src/plans/conversion/type_coercion/binary.rs [hunk 2]
```diff
@@ -115,6 +115,21 @@ fn err_date_str_compare() -> PolarsResult<()> {
     }
 }
 
+#[cfg(feature = "dtype-duration")]
+fn err_duration_str_compare() -> PolarsResult<()> {
```
Comment (nameexhaustion) [inline crates/polars-plan/src/plans/conversion/type_coercion/binary.rs:119]: This doesn't seem related to this PR, could we remove it?

File: crates/polars-plan/src/plans/conversion/type_coercion/binary.rs [hunk 3]
```diff
@@ -256,6 +271,13 @@ pub(super) fn process_binary(
         (Time | Unknown(UnknownKind::Str), String, op) if op.is_comparison_or_bitwise() => {
             err_date_str_compare()?
         },
+        #[cfg(feature = "dtype-duration")]
+        (Duration(_), String | Unknown(UnknownKind::Str), op)
+        | (String | Unknown(UnknownKind::Str), Duration(_), op)
+            if op.is_comparison_or_bitwise() =>
```
Comment (nameexhaustion) [inline crates/polars-plan/src/plans/conversion/type_coercion/binary.rs:277]: This doesn't seem related to this PR, could we remove it?

File: crates/polars-plan/src/plans/conversion/type_coercion/binary.rs (hunk 2 above)
Comment (kanenorman) [reply crates/polars-plan/src/plans/conversion/type_coercion/binary.rs:119]: I made this change to address [`test_comparison_with_string_raises_9461`](https://github.com/pola-rs/polars/blob/3ee3f1baa8deedc712e45d822cafc62eef2a3af7/py-polars/tests/unit/datatypes/test_duration.py#L221).

Without it, `df.filter(pl.col("duration") > "1h")` wouldn't raise an `InvalidOperationError`, which breaks the current behavior where comparing `Duration` to a `String` should trigger an error.

For reference there's a similar check for `DateTime` in [err_date_str_compare](https://github.com/pola-rs/polars/blob/3ee3f1baa8deedc712e45d822cafc62eef2a3af7/crates/polars-plan/src/plans/conversion/type_coercion/binary.rs#L104).

What do you think the expected behavior should be for `Duration`/`String` comparisons?
Should we raise an error, or is there another approach you'd prefer?

File: crates/polars-plan/src/plans/conversion/type_coercion/binary.rs (hunk 3 above)
Comment (kanenorman) [reply crates/polars-plan/src/plans/conversion/type_coercion/binary.rs:277]: same context as [above](https://github.com/pola-rs/polars/pull/26860/changes/BASE..8823f5fe5567cf0c3b3d726d6e1ab7331e7fb84c#r2935953177) 😊

File: crates/polars-core/src/series/implementations/duration.rs (hunk 1 above)
Comment (kanenorman) [reply crates/polars-core/src/series/implementations/duration.rs]: agree. will do 😊

File: crates/polars-plan/src/plans/conversion/type_coercion/binary.rs (hunk 2 above)
Comment (nameexhaustion) [reply crates/polars-plan/src/plans/conversion/type_coercion/binary.rs:119]: Alright, this code is needed then, let's keep it

**Comment by kanenorman** [comment]
Added test for LazyFrame and an additional test for direct comparison between polars `Duration` and `datetime.timedelta` since the new error message instructs users to explicitly compare using `timedelta` instances rather than `String`.

@nameexhaustion

**Comment by kanenorman** [comment]
@nameexhaustion - is this good to merge?
